ReadJournal query "currentPersistenceIds" with Casbah leads to MongoCommandException if response is > 16MB

Hi.
I am using the mentioned ReadJournal query `currentPersistenceIds` for Casbah and have > 200.000 persistenceIds in the Database.
Apparently those > 200.000 ids form a Document > 16MB and I get following Exception:
```
2016-02-18T14:29:19.000+00:00 [App/0] ERR Caused by: com.mongodb.MongoCommandException: Command failed with error 17217: 'exception: distinct too big, 16mb cap' on server mdb001uans1.clouduser.bicadmin.com:30000. The full response is { "errmsg" : "exception: distinct too big, 16mb cap", "code" : 17217, "ok" : 0.0 }
2016-02-18T14:29:19.000+00:00 [App/0] ERR at com.mongodb.connection.ProtocolHelper.getCommandFailureException(ProtocolHelper.java:77) ~[app/:na]
2016-02-18T14:29:19.000+00:00 [App/0] ERR at com.mongodb.connection.CommandProtocol.createCommandResult(CommandProtocol.java:140) ~[app/:na]
2016-02-18T14:29:19.000+00:00 [App/0] ERR at com.mongodb.connection.CommandProtocol.receiveMessage(CommandProtocol.java:131) ~[app/:na]
2016-02-18T14:29:19.000+00:00 [App/0] ERR at com.mongodb.connection.CommandProtocol.execute(CommandProtocol.java:84) ~[app/:na]
2016-02-18T14:29:19.000+00:00 [App/0] ERR at com.mongodb.connection.DefaultServer$DefaultServerProtocolExecutor.execute(DefaultServer.java:155) ~[app/:na]
2016-02-18T14:29:19.000+00:00 [App/0] ERR at com.mongodb.connection.DefaultServerConnection.executeProtocol(DefaultServerConnection.java:219) ~[app/:na]
```

--- a/casbah/src/main/scala/akka/contrib/persistence/mongodb/CasbahPersistenceReadJournaller.scala
+++ b/casbah/src/main/scala/akka/contrib/persistence/mongodb/CasbahPersistenceReadJournaller.scala
@@ -5,6 +5,7 @@ import com.mongodb.casbah.Imports._
 import com.mongodb.{Bytes, DBObject}
 
 import scala.concurrent.Future
+import scala.util.Random
 
 object CurrentAllPersistenceIds {
   def props(driver: CasbahMongoDriver): Props = Props(new CurrentAllPersistenceIds(driver))
@@ -13,11 +14,17 @@ object CurrentAllPersistenceIds {
 class CurrentAllPersistenceIds(val driver: CasbahMongoDriver) extends SyncActorPublisher[String, Stream[String]] {
   import CasbahSerializers._
 
-  override protected def initialCursor: Stream[String] =
-    driver.journal
-          .distinct(PROCESSOR_ID, MongoDBObject())
-          .toStream
-          .collect { case s:String => s }
+  val temporaryCollectionName = s"persistenceids-${System.currentTimeMillis()}-${Random.nextInt(1000)}"
+
+  override protected def initialCursor: Stream[String] = {
+    driver.journal.aggregate(
+      MongoDBObject("$project" -> MongoDBObject(PROCESSOR_ID -> 1)) ::
+        MongoDBObject("$group" -> MongoDBObject("_id" -> s"$$$PROCESSOR_ID")) ::
+        MongoDBObject("$out" -> temporaryCollectionName) ::
+        Nil).results
+
+    driver.collection(temporaryCollectionName).find().toStream.flatMap(_.getAs[String]("_id"))
+  }
 
   override protected def next(c: Stream[String], atMost: Long): (Vector[String], Stream[String]) = {
     val (buf,remainder) = c.splitAt(atMost.toIntWithoutWrapping)
@@ -28,7 +35,9 @@ class CurrentAllPersistenceIds(val driver: CasbahMongoDriver) extends SyncActorP
     c.isEmpty
   }
 
-  override protected def discard(c: Stream[String]): Unit = ()
+  override protected def discard(c: Stream[String]): Unit = {
+    driver.collection(temporaryCollectionName).drop()
+  }
 }
 
 object CurrentAllEvents {

--- a/casbah/src/main/scala/akka/contrib/persistence/mongodb/CasbahPersistenceSnapshotter.scala
+++ b/casbah/src/main/scala/akka/contrib/persistence/mongodb/CasbahPersistenceSnapshotter.scala
@@ -14,9 +14,11 @@ object CasbahPersistenceSnapshotter {
   import SnapshottingFieldNames._
 
   implicit def serializeSnapshot(snapshot: SelectedSnapshot)(implicit serialization: Serialization): DBObject = {
-    val obj = MongoDBObject(PROCESSOR_ID -> snapshot.metadata.persistenceId,
+    val obj = MongoDBObject(
+      PROCESSOR_ID -> snapshot.metadata.persistenceId,
       SEQUENCE_NUMBER -> snapshot.metadata.sequenceNr,
-      TIMESTAMP -> snapshot.metadata.timestamp)
+      TIMESTAMP -> snapshot.metadata.timestamp
+    )
     snapshot.snapshot match {
       case o: DBObject =>
         obj.put(V2.SERIALIZED, o)
@@ -80,7 +82,10 @@ class CasbahPersistenceSnapshotter(driver: CasbahMongoDriver) extends MongoPersi
   }
 
   private[mongodb] def saveSnapshot(snapshot: SelectedSnapshot)(implicit ec: ExecutionContext) = Future {
-    snaps.insert(snapshot, writeConcern)
+    val query = MongoDBObject(PROCESSOR_ID -> snapshot.metadata.persistenceId,
+                              SEQUENCE_NUMBER -> snapshot.metadata.sequenceNr,
+                              TIMESTAMP -> snapshot.metadata.timestamp)
+    snaps.update(query, snapshot, upsert = true, multi = false, writeConcern)
     ()
   }
 

--- a/common/src/main/scala/akka/contrib/persistence/mongodb/MongoReadJournal.scala
+++ b/common/src/main/scala/akka/contrib/persistence/mongodb/MongoReadJournal.scala
@@ -195,8 +195,10 @@ trait SyncActorPublisher[A,Cursor] extends ActorPublisher[A] {
     case Request(_) =>
       val (filled,remaining) = next(cursor, totalDemand)
       filled foreach onNext
-      if (isCompleted(remaining))
+      if (isCompleted(remaining)) {
         onCompleteThenStop()
+        discard(remaining)
+      }
       else
         context.become(streaming(remaining, offset + filled.size))
   }

--- a/common/src/test/scala/akka/contrib/persistence/mongodb/ReadJournalSpec.scala
+++ b/common/src/test/scala/akka/contrib/persistence/mongodb/ReadJournalSpec.scala
@@ -193,7 +193,7 @@ abstract class ReadJournalSpec[A <: MongoPersistenceExtension](extensionClass: C
     eventually {
       mongoClient.getDatabase(embedDB).listCollectionNames()
         .into(new java.util.HashSet[String]()).asScala.filter(_.startsWith("persistenceids-")) should have size 0L
-    }(PatienceConfig(timeout = Span(1L, Seconds), interval = Span(5L, Seconds)))
+    }(PatienceConfig(timeout = Span(5L, Seconds), interval = Span(500L, Millis)))
   }
 
   it should "support the all persistence ids query" in withConfig(config(extensionClass), "akka-contrib-mongodb-persistence-readjournal") { case (as,_)  =>

--- a/common/src/test/scala/akka/contrib/persistence/mongodb/ReadJournalSpec.scala
+++ b/common/src/test/scala/akka/contrib/persistence/mongodb/ReadJournalSpec.scala
@@ -3,22 +3,30 @@ package akka.contrib.persistence.mongodb
 import akka.actor.Props
 import akka.persistence.PersistentActor
 import akka.persistence.query.{EventEnvelope, PersistenceQuery}
-import akka.stream.{ClosedShape, ActorMaterializer}
-import akka.stream.scaladsl.{RunnableGraph, Sink, Merge, GraphDSL}
+import akka.stream.scaladsl.{GraphDSL, Merge, RunnableGraph, Sink}
+import akka.stream.{ActorMaterializer, ClosedShape}
 import akka.testkit._
+import com.mongodb.client.model.{BulkWriteOptions, InsertOneModel}
 import com.typesafe.config.ConfigFactory
-import org.scalatest.BeforeAndAfterAll
+import org.bson.Document
 import org.scalatest.concurrent.Eventually
+import org.scalatest.time._
+import org.scalatest.{BeforeAndAfter, BeforeAndAfterAll}
 
 import scala.concurrent.{Await, Future, Promise}
+import scala.util.Random
 
-abstract class ReadJournalSpec[A <: MongoPersistenceExtension](extensionClass: Class[A], dbName: String) extends BaseUnitTest with ContainerMongo with BeforeAndAfterAll with Eventually {
+abstract class ReadJournalSpec[A <: MongoPersistenceExtension](extensionClass: Class[A], dbName: String) extends BaseUnitTest with ContainerMongo with BeforeAndAfterAll with BeforeAndAfter with Eventually {
 
   import ConfigLoanFixture._
 
   override def embedDB = s"read-journal-test-$dbName"
 
   override def afterAll() = cleanup()
+
+  before {
+    "akka_persistence_realtime" :: "akka_persistence_journal" :: Nil foreach(mongoClient.getDatabase(embedDB).getCollection(_).drop())
+  }
 
   def config(extensionClass: Class[_]) = ConfigFactory.parseString(s"""
     |akka.contrib.persistence.mongodb.mongo.driver = "${extensionClass.getName}"
@@ -141,10 +149,51 @@ abstract class ReadJournalSpec[A <: MongoPersistenceExtension](extensionClass: C
 
     val fut = readJournal.currentPersistenceIds().runFold(Seq.empty[String])(_ :+ _)
 
-    import collection.JavaConverters._
-    val pids = mongoClient.getDB(embedDB).getCollection("akka_persistence_journal").distinct("pid").asScala.toList.map(_.toString)
-
     Await.result(fut,10.seconds.dilated) should contain allOf("1","2","3","4","5")
+  }
+
+  it should "support the current persistence ids query with more than 16MB of ids" in withConfig(config(extensionClass), "akka-contrib-mongodb-persistence-readjournal") { case (as, _) =>
+    import concurrent.duration._
+
+    implicit val system = as
+    implicit val ec = as.dispatcher
+    implicit val mat = ActorMaterializer()
+
+    val alphabet = "abcdefghijklmnopqrstuvwxyz0123456789"
+    val PID_SIZE = 900
+    val EVENT_COUNT = 187 * 100
+
+    PID_SIZE * EVENT_COUNT shouldBe > (16 * 1024 * 1024)
+
+    def pidGen = (0 to PID_SIZE).map(_ => alphabet.charAt(Random.nextInt(alphabet.length))).mkString
+
+    import collection.JavaConverters._
+
+    val journalCollection = mongoClient.getDatabase(embedDB).getCollection("akka_persistence_journal")
+    Stream.from(1).takeWhile(_ <= EVENT_COUNT).map{i =>
+      new Document()
+        .append(JournallingFieldNames.PROCESSOR_ID, s"$pidGen-$i")
+          .append(JournallingFieldNames.FROM, 0L)
+          .append(JournallingFieldNames.TO, 0L)
+          .append(JournallingFieldNames.VERSION, 1)
+    }.grouped(1000).foreach{ dbos =>
+      val batch = dbos.toList.map(new InsertOneModel(_)).asJava
+      journalCollection.bulkWrite(batch, new BulkWriteOptions().ordered(false).bypassDocumentValidation(true))
+      ()
+    }
+
+    mongoClient.getDatabase(embedDB).getCollection("akka_persistence_journal").count() shouldBe EVENT_COUNT
+
+    val readJournal =
+      PersistenceQuery(as).readJournalFor[ScalaDslMongoReadJournal](MongoReadJournal.Identifier)
+
+    val fut = readJournal.currentPersistenceIds().runFold(0){ case (inc,_) => inc + 1}
+    Await.result(fut,10.seconds.dilated) shouldBe EVENT_COUNT
+
+    eventually {
+      mongoClient.getDatabase(embedDB).listCollectionNames()
+        .into(new java.util.HashSet[String]()).asScala.filter(_.startsWith("persistenceids-")) should have size 0L
+    }(PatienceConfig(timeout = Span(1L, Seconds), interval = Span(5L, Seconds)))
   }
 
   it should "support the all persistence ids query" in withConfig(config(extensionClass), "akka-contrib-mongodb-persistence-readjournal") { case (as,_)  =>

--- a/rxmongo/src/main/scala/akka/contrib/persistence/mongodb/RxMongoReadJournaller.scala
+++ b/rxmongo/src/main/scala/akka/contrib/persistence/mongodb/RxMongoReadJournaller.scala
@@ -3,10 +3,13 @@ package akka.contrib.persistence.mongodb
 import akka.actor._
 import akka.contrib.persistence.mongodb.JournallingFieldNames._
 import akka.stream.actor.ActorPublisher
+import akka.{Done => ADone}
 import play.api.libs.iteratee._
-import reactivemongo.api.commands.Command
-import reactivemongo.api.{BSONSerializationPack, QueryOpts}
+import reactivemongo.api.QueryOpts
 import reactivemongo.bson._
+
+import scala.concurrent.{ExecutionContext, Future}
+import scala.util.Random
 
 trait IterateeActorPublisher[T] extends ActorPublisher[T] with ActorLogging {
 
@@ -24,39 +27,55 @@ trait IterateeActorPublisher[T] extends ActorPublisher[T] with ActorLogging {
 
   private case class Next(enumerator: Enumerator[T], iteratee: Iteratee[T,Unit])
 
-  private var completed: Boolean = false
-
   private def respondToDemand(enumerator: Enumerator[T], iter: Iteratee[T,Unit]) = {
-    enumerator(iter).map(Next(enumerator,_)).pipeTo(self)
+    Concurrent.runPartial(enumerator, iter).map{ case(_,e) => Next(e,iter)}.pipeTo(self)
   }
 
-  val nextElem: Iteratee[T,Unit] = {
-    Cont {
-      case Input.EOF =>
-        onComplete()
-        completed = true
-        Done(())
-      case Input.Empty =>
-        nextElem
-      case Input.El(elem) =>
-        onNext(elem)
-        nextElem
+  private class DoneIteratee extends Iteratee[T, Unit] {
+    override def fold[B](folder: (Step[T,Unit]) => Future[B])(implicit ec: ExecutionContext): Future[B] = {
+      folder(Step.Done((),Input.Empty))
     }
   }
 
+  private class CursorIteratee extends Iteratee[T, Unit] {
+    override def fold[B](folder: (Step[T, Unit]) => Future[B])(implicit ec: ExecutionContext): Future[B] = {
+      folder(Step.Cont({
+        case Input.El(elem) =>
+          onNext(elem)
+          if (totalDemand > 0L) this
+          else new DoneIteratee
+        case Input.EOF =>
+          onComplete()
+          cleanup().pipeTo(self)
+          new DoneIteratee
+        case Input.Empty =>
+          if (totalDemand > 0L) this
+          else new DoneIteratee
+      }))
+    }
+  }
+
+  def cleanup(): Future[ADone] = Future.successful(ADone)
+
   def defaults: Receive = {
     case _:Cancel|SubscriptionTimeoutExceeded =>
-      log.debug(s"Cancelling stream")
+      log.warning(s"Cancelling stream")
       onCompleteThenStop()
+      cleanup().pipeTo(self)
+      ()
     case Status.Failure(t) =>
       log.error(t,"Failure occurred while streaming")
       onErrorThenStop(t)
+      cleanup().pipeTo(self)
+      ()
+    case ADone =>
+      context.stop(self)
   }
 
   def awaiting(enumerator: Enumerator[T]): Receive = defaults orElse handleNext("Awaiting") orElse {
     case Request(_) =>
       log.debug(s"Awaiting: Request received, demand = $totalDemand")
-      respondToDemand(enumerator, nextElem)
+      respondToDemand(enumerator, new CursorIteratee)
       context.become(streaming(enumerator))
   }
 
@@ -65,11 +84,7 @@ trait IterateeActorPublisher[T] extends ActorPublisher[T] with ActorLogging {
   private def handleNext(header: String): Receive = {
     case Next(enm, it) =>
       log.debug("Next received")
-      if (completed) {
-        log.debug(s"$header: completed - now stopping")
-        context.stop(self)
-      }
-      else if (totalDemand > 0) {
+      if (totalDemand > 0) {
         log.debug(s"$header: requesting more, demand = $totalDemand")
         respondToDemand(enm, it)
         context.become(streaming(enm))
@@ -118,22 +133,30 @@ class CurrentAllPersistenceIds(val driver: RxMongoDriver) extends IterateeActorP
   import context.dispatcher
   import reactivemongo.bson._
 
-  private def flatten(doc: BSONDocument): Enumerator[String] = {
-    val result = for {
-      arr <- doc.getAs[BSONArray]("values").toStream
-      elem <- arr.values
-      bson <- elem.seeAsOpt[BSONString].toStream
-    } yield bson.value
-    Enumerator.enumerate[String](result)
+  val temporaryCollectionName = s"persistenceids-${System.currentTimeMillis()}-${Random.nextInt(1000)}"
+
+  override def cleanup() = {
+    driver.collection(temporaryCollectionName).drop().map(_ => ADone)
   }
 
+  val flattened = Enumeratee.mapConcat[BSONDocument](_.getAs[String]("_id").toSeq)
+
   override def initial = {
-    val q = BSONDocument("distinct" -> driver.journalCollectionName, "key" -> PROCESSOR_ID, "query" -> BSONDocument())
-    val cmd = Command.run(BSONSerializationPack)
-    cmd(driver.db,cmd.rawCommand(q))
-      .cursor[BSONDocument]
-      .enumerate()
-      .flatMap(flatten)
+    import driver.journal.BatchCommands.AggregationFramework.{Group, Out, Project}
+
+    val enumerator = for {
+     _ <- driver.journal.aggregate(Project(BSONDocument(PROCESSOR_ID -> 1)),
+                                    List(
+                                      Group(BSONString(s"$$$PROCESSOR_ID"))(),
+                                      Out(temporaryCollectionName)
+                                    ))
+     results <- Future.successful(driver.collection(temporaryCollectionName)
+                                        .find(BSONDocument())
+                                        .cursor[BSONDocument]()
+                                        .enumerate().through(flattened))
+    } yield results
+
+    Enumerator.flatten(enumerator)
   }
 }
 

--- a/rxmongo/src/main/scala/akka/contrib/persistence/mongodb/RxMongoSnapshotter.scala
+++ b/rxmongo/src/main/scala/akka/contrib/persistence/mongodb/RxMongoSnapshotter.scala
@@ -27,8 +27,14 @@ class RxMongoSnapshotter(driver: RxMongoDriver) extends MongoPersistenceSnapshot
     selected
   }
 
-  private[mongodb] def saveSnapshot(snapshot: SelectedSnapshot)(implicit ec: ExecutionContext) =
-    snaps.insert(snapshot, writeConcern).map(_ => ())
+  private[mongodb] def saveSnapshot(snapshot: SelectedSnapshot)(implicit ec: ExecutionContext) = {
+    val query = BSONDocument(
+      PROCESSOR_ID -> snapshot.metadata.persistenceId,
+      SEQUENCE_NUMBER -> snapshot.metadata.sequenceNr,
+      TIMESTAMP -> snapshot.metadata.timestamp
+    )
+    snaps.update(query, snapshot, writeConcern, upsert = true, multi = false).map(_ => ())
+  }
 
   private[mongodb] def deleteSnapshot(pid: String, seq: Long, ts: Long)(implicit ec: ExecutionContext) = {
     val criteria =

